File: src/main/resources/ts/page/calendar.ts
import { assembly } from "..";

export async function calendar() {
    const json = await (await fetch("/service/calendar", { method: "POST" })).json() as Calendar;
    const date = new Date();
    const year = date.getFullYear();
    let su_month = date.getMonth()+1;

    (window as any).nowDay = date.getDate();
    (window as any).now_month = su_month;
    (window as any).swap_calendar = swap_calendar;
    (document.getElementById("calendarTable_"+su_month) as HTMLElement).innerHTML = getCalendar(su_month);

    function getCalendar(month:number) {
        (window as any).lastDay = new Date(year, month, 0).getDate();
        (window as any).fristWeek = new Date(year, month-1, 1).getDay()+1;
        (window as any).month = month;
        assembly.init(2);
        
        json.SchoolSchedule[1].row.forEach((row) => {
            const month_str = row.AA_YMD.substring(4, 6);
            const day = row.AA_YMD.substring(6, 8);
            
            if(Number(month_str) == month) {
                (window as any).row = row;
                const date = new Date(year +"-"+ month_str +"-"+ day);
                assembly.calendar_push(date.getDate(), month);
            }
        });

        return assembly.getCalendar();
    }

    function swap_calendar(month:string|number) {
        const month_nmb = Number(month);
        const calendar_table = document.getElementById("calendarTable_"+month) as HTMLElement;
        if(calendar_table.innerHTML == '') {
            calendar_table.innerHTML = getCalendar(month_nmb);
        }
        (document.getElementById("calendar_"+month) as HTMLElement).style.display = 'block';

        (document.getElementById("calendar_"+su_month) as HTMLElement).style.display = 'none';
        su_month = month_nmb
    }
}




type GRADE_EVENT = "Y" | "N" | "*";

type row = {
    ATPT_OFCDC_SC_CODE:string,
    SD_SCHUL_CODE:string,
    AY:string,
    AA_YMD:string, // 학사일자
    ATPT_OFCDC_SC_NM:string,
    SCHUL_NM:string,
    DGHT_CRSE_SC_NM:string,
    SCHUL_CRSE_SC_NM:string,
    EVENT_NM:string, // 행사명
    EVENT_CNTNT:string, // 행사내용
    ONE_GRADE_EVENT_YN: GRADE_EVENT, // 1학년행사여부
    TW_GRADE_EVENT_YN: GRADE_EVENT, // 2학년행사여부
    THREE_GRADE_EVENT_YN: GRADE_EVENT, // 3학년행사여부
    FR_GRADE_EVENT_YN: GRADE_EVENT, // 4학년행사여부
    FIV_GRADE_EVENT_YN: GRADE_EVENT, // 5학년행사여부
    SIX_GRADE_EVENT_YN: GRADE_EVENT, // 6학년행사여부
    SBTR_DD_SC_NM: "공휴일" | "휴업일" | "해당없음",
    LOAD_DTM: string
}

type Calendar = {
    SchoolSchedule: [
        {
            head: [
                {
                    list_total_count:number
                },
                {
                    RESULT: {
                        CODE:string,
                        MESSAGE:string
                    }
                }
            ]
        },
        {
            row:row[]
        }
    ]
}

File: src/main/resources/cpp/page/calendar.cpp
#include "calendar.h"

Calendar::Calendar() {
    window = val::global("window");
    int lastDay = window["lastDay"].as<int>();
    now_month = window["now_month"].as<int>();
    fristWeek = window["fristWeek"].as<int>();
    nowDay = window["nowDay"].as<int>();

    int day = 1, week = fristWeek;
    for(int i=0; i < 5; i++) {
        for(int j=week; j <= 7; j++) {
            string className = "";
            string classTd = "";
            string EVENT_NM = "평일";
        
            if(j == 7 || j == 1) {
                className = "holiday";
                EVENT_NM = "주말";
            }

            if(nowDay == day) {
                classTd = "current-day";
            }

            calendar_table[i][j] = "<td class='"+classTd+"'>\
                <div class='day'>\
                    <p class='day_p "+className+"'>"+to_string(day)+"</p>"+
                    EVENT_NM+
                "</div>\
            </td>";
            day++;

            if(day > lastDay) return;

            week = 1;
        }
    }
}

Calendar::~Calendar() {};

void Calendar::run(int date, int month) {
    val row = window["row"];
    int week = fristWeek, day = 0, is = 0;
    string classTd, className, EVENT_NM, EVENT_CNTNT, SBTR_DD_SC_NM;
    string* table = NULL;

    for(int i=0; i < 5; i++) {
        for(int j=week; j <= 7; j++) {
            table = &calendar_table[i][j];
            if(*table == "<td></td>") continue;

            day++;
            if(day == nowDay && month == now_month) {
                if(table->find("<br>") != string::npos) continue;
                is = 2;
                classTd = "current-day";
            }else if(day == date) {
                is = 1;
                classTd = "";
            }

            if(is) {
                if(day == date) {
                    EVENT_NM = row["EVENT_NM"].as<string>();
                    EVENT_CNTNT = row["EVENT_CNTNT"].as<string>();
                    SBTR_DD_SC_NM = row["SBTR_DD_SC_NM"].as<string>();
                    if(is == 2) classTd = "current-day";
                }else {
                    EVENT_NM = "";
                    EVENT_CNTNT = "";
                    SBTR_DD_SC_NM = "해당없음";
                }

                if(SBTR_DD_SC_NM != "해당없음") {
                    className = "holiday";
                }
                
                if(EVENT_NM == "토요휴업일" || j == 7 || j == 1) {
                    className = "holiday";
                    EVENT_NM = "주말";
                }

                *table = "<td class='"+classTd+"'>\
                    <div class='day'>\
                        <p class='day_p "+className+"'>"+to_string(day)+"</p>"+
                        EVENT_NM+"<br>"+EVENT_CNTNT+
                    "</div>\
                </td>";

                if(is != 2) return;
                week = 1;
            }else {
                week = 1;
            }
            is = 0;
        }
    }
    cout << "에러 (날짜를 찾지 못함)" << endl;
}

char* Calendar::get() {
    string result = "";
    for(int i=0; i < 5; i++) {
        for(int j=0; j < 9; j++) {
            result.append(calendar_table[i][j]);
        }
    }
    char* str = const_cast<char*>(result.c_str());
    return str;
}